pkg/cmd/operator/cmd.go

```diff
@@ -99,6 +105,70 @@ func NewOperator() *cobra.Command {
+					log.WithError(err).Fatal("failed to parse kubeconfig")
+				}
+				coreClient, err := corev1client.NewForConfig(cfg)
+				if err != nil {
+					log.WithError(err).Fatal("failed to set up client")
+				}
+
+				// The status quo for this controller is to LIST + WATCH all Secrets on the cluster. This consumes
+				// more resources than necessary on clusters where users put other data in Secrets themselves, as we
+				// hold that data in our cache and never do anything with it. The reconcilers mainly need to react to
+				// changes in Secrets created for CredentialRequests, which they control and can label, allowing us
+				// to filter the LIST + WATCH down and hold the minimal set of data in memory. However, two caveats:
+				// - in passthrough and mint mode, admin credentials are also provided by the user through Secrets,
+				//   and we need to watch those, but we can't label them
+				// - on existing clusters, secrets exist from previous work these reconcilers did that will not have
+				//   Secrets labelled
+				// We could solve the second issue with an interim release of this controller that labels all previous
+				// Secrets, but does not restrict the watch stream.
+				// Due to the way that controller-runtime closes over the client/cache concepts, it's difficult to
+				// solve the first issue, though, since we'd need two sets of clients and caches, both for Secrets,
+				// and ensure that we use one for client access to Secrets we're creating or mutating and the other
+				// when we're interacting with admin credentials. Not impossible to do, but tricky to implement and
+				// complex.
+				// Until we undertake that effort, we apply a simplification to the space: only when AWS STS mode is
+				// enabled, we will try to filter the LIST + WATCH. This mode is brand new, so we can be reasonably
+				// sure that there are no previous secrets on the cluster, and, we make the filtering best-effort
+				// in order to check if that assumption held. Second, AWS STS mode only runs in clusters without
+				// admin credentials, so if we apply the filter, we should not see failures downstream from clients
+				// that hope to see those objects but can't.
+				var filteredWatchPossible bool
+				if awsSecurityTokenServiceGateEnabled {
+					secrets, err := coreClient.Secrets(metav1.NamespaceAll).List(ctx, metav1.ListOptions{})
+					if err != nil {
+						log.WithError(err).Fatal("failed to list secrets")
+					}
+					var missing []types.NamespacedName
+					for _, secret := range secrets.Items {
+						if credentialsrequest.IsMissingSecretLabel(&secret) {
+							missing = append(missing, types.NamespacedName{
+								Namespace: secret.Namespace,
+								Name:      secret.Name,
+							})
+						}
+					}
+					if len(missing) != 0 {
+						log.WithField("missing", missing).Warnf("%s feature gate enabled but not all secrets labelled, falling back to caching all secrets on cluster", configv1.FeatureGateAWSSecurityTokenService)
```

Comment:
thought you were going to attempt to add labels to secrets that are missing them and only if that failed, fallback to caching all secrets?

Reply:
In the controller that runs, not at startup.

Reply:
I could not find a clean way to thread the context for  the controller down far enough to where we could cancel it and cause self-destruction/restart. Happy to put a `log.Fatal()` in there but it seems like making that control flow abnormal might be hard to reason about for other devs in the future.